File: binary search/34. Find First and Last Position of Element in Sorted Array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int[] res = new int[2];
        res[0] = search(nums, target, true);
        res[1] = search(nums, target, false);
        return res;
    }

    public int search(int[] nums, int target, boolean findFirst){
        int l = 0;
        int r = nums.length - 1;
        int ans = -1;
        while(l <= r){
            int mid = l + (r - l) / 2;
            if(nums[mid] == target){
                ans = mid;
                if(findFirst) r = mid - 1;
                else l = mid + 1;
            } else {
                if(nums[mid] < target) l = mid + 1;
                else r = mid - 1;
            }
        }
        return ans;
    }
}

File: binary search/34. Find First and Last Position of Element in Sorted Array/34. Find First and Last Position of Element in Sorted Array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function (nums, target) {
  let l = 0,
    r = nums.length - 1,
    mid;
  while (l <= r) {
    mid = l + ((r - l) >> 1);
    if (nums[mid] === target) {
      let low = mid,
        high = mid;
      while (low >= 0 && nums[low] === target) {
        low--;
      }
      while (high < nums.length && nums[high] === target) {
        high++;
      }
      return [low + 1, high - 1];
    } else {
      nums[mid] > target ? (r = mid - 1) : (l = mid + 1);
    }
  }
  return [-1, -1];
};
